### PR TITLE
Various RPC fixes

### PR DESCRIPTION
**Pull request by ebuchman**
Fixes:

- https://github.com/tendermint/tendermint/issues/1481
- https://github.com/tendermint/tendermint/issues/1480
- https://github.com/tendermint/tendermint/issues/1310
- https://github.com/tendermint/tendermint/issues/1199 (?)

### CODE REVIEW

File: node/node.go [hunk 1]
```diff
@@ -412,6 +402,16 @@ func (n *Node) OnStart() error {
 	// Add ourselves to addrbook to prevent dialing ourselves
 	n.addrBook.AddOurAddress(nodeInfo.NetAddress())
 
+	// Run the RPC server first
+	// so we can eg. receive txs for the first block
```
Comment (melekes) [inline node/node.go]: first in relation to what?

File: types/part_set.go [hunk 2]
```diff
@@ -264,3 +264,20 @@ func (ps *PartSet) StringShort() string {
 	defer ps.mtx.Unlock()
 	return fmt.Sprintf("(%v of %v)", ps.Count(), ps.Total())
 }
+
+func (ps *PartSet) MarshalJSON() ([]byte, error) {
+	if ps == nil {
+		return []byte("nil-PartSet"), nil
```
Comment (melekes) [inline types/part_set.go]: shouldn't this be a valid json? like `null`? or maybe `[]byte("\"nil-PartSet\"")`

File: types/part_set.go [hunk 3]
```diff
@@ -264,3 +264,20 @@ func (ps *PartSet) StringShort() string {
...
+	}
+
+	ps.mtx.Lock()
+	defer ps.mtx.Unlock()
+
+	return cdc.MarshalJSON(struct {
+		CountTotal    string        `json:"count/total"`
```
Comment (melekes) [inline types/part_set.go]: why not make them 2 separate fields?

File: types/vote_set.go [hunk 4]
```diff
@@ -445,6 +445,26 @@ func (voteSet *VoteSet) StringIndented(indent string) string {
 		indent)
 }
 
+// Marshal the VoteSet to JSON. Same as String(), just in JSON,
+// and without the height/round/type_ (since its already included in the votes).
+func (voteSet *VoteSet) MarshalJSON() ([]byte, error) {
+	voteStrings := make([]string, len(voteSet.votes))
+	for i, vote := range voteSet.votes {
```
Comment (melekes) [inline types/vote_set.go]: what about 

```
voteSet.mtx.Lock()
defer voteSet.mtx.Unlock()
```

File: types/part_set.go [hunk 5]
```diff
@@ -264,3 +264,20 @@ func (ps *PartSet) StringShort() string {
...
+		PartsBitArray *cmn.BitArray `json:"parts_bit_array"`
+	}{
+		fmt.Sprintf("%d/%d", ps.Count(), ps.Total()),
+		ps.partsBitArray.Copy(),
```
Comment (melekes) [inline types/part_set.go]: now that we have custom marshaller for BitArray's, I think we can drop `.Copy()` https://github.com/tendermint/tmlibs/blob/develop/common/bit_array.go#L327

File: node/node.go (hunk 1 above)
Comment (ebuchman) [reply node/node.go]: to the p2p server we start after

File: types/part_set.go (hunk 2 above)
Comment (ebuchman) [reply types/part_set.go]: yes it should. maybe we should just do `{}`

File: types/part_set.go (hunk 3 above)
Comment (ebuchman) [reply types/part_set.go]: Figured there's no need to take up the extra space but maybe that's silly (dump consensus is a big dump already). Also we're typically more interested in the ratio than we are in the two numbers

File: types/part_set.go (hunk 5 above)
Comment (ebuchman) [reply types/part_set.go]: ++

File: types/vote_set.go (hunk 4 above)
Comment (ebuchman) [reply types/vote_set.go]: probably right. String() doesn't have it tho

**Comment by ebuchman** [comment]
Thanks for the review @melekes !

**Pull request merged**